src/Illuminate/Database/Eloquent/Concerns/HasCasts.php

```diff
@@ -0,0 +1,360 @@
+     * @param  string  $cast
+     * @return bool
+     */
+    protected function isCustomDateTimeCast($cast)
+    {
+        return strncmp($cast, 'date:', 5) === 0 ||
+            strncmp($cast, 'datetime:', 9) === 0;
+    }
+
+    /**
+     * Determine if the cast type is a decimal cast.
+     *
+     * @param  string  $cast
+     * @return bool
+     */
+    protected function isDecimalCast($cast)
+    {
+        return strncmp($cast, 'decimal:', 8) === 0;
+    }
+
+    /**
+     * Determine if the given attribute is a date or date castable.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isDateAttribute($key)
+    {
+        return in_array($key, $this->getDates(), true) ||
+            $this->isDateCastable($key);
+    }
+
+    /**
+     * Is the checked value a custom Cast.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isCustomCastable($key)
+    {
+        if ($cast = $this->getCast($key)) {
+            return is_subclass_of($cast, Castable::class);
+        }
+
+        return false;
+    }
+
+    /**
+     * Determine whether a value is Date / DateTime castable for inbound manipulation.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isDateCastable($key)
+    {
+        return $this->hasCast($key, ['date', 'datetime']);
+    }
+
+    /**
+     * Determine whether a value is JSON castable for inbound manipulation.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isJsonCastable($key)
+    {
+        return $this->hasCast($key, ['array', 'json', 'object', 'collection']);
+    }
+
+    /**
+     * Getting the execution result from a user Cast object.
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     * @return mixed
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function fromCustomCastable($key, $value = null)
+    {
+        return $this
+            ->normalizeHandlerToCallable($key)
+            ->get($value);
+    }
+
+    /**
+     * Converting a value by custom Cast.
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     * @return mixed
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function toCustomCastable($key, $value = null)
+    {
+        return $this
+            ->normalizeHandlerToCallable($key)
+            ->set($value);
+    }
+
+    /**
+     * Getting a custom cast instance.
+     *
+     * @param  string  $key
+     * @return \Illuminate\Contracts\Database\Eloquent\Castable
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function normalizeHandlerToCallable($key)
+    {
+        if (! array_key_exists($key, $this->casts_instances)) {
+            $cast = $this->getCast($key);
+
+            $this->casts_instances[$key] = new $cast;
```

Comment:
1. Regarding memory consumption, would it be a good idea to keep the cast instances in a static array on the model? So the resolved casts are cached between model instances? Something like what is done to the global scopes.
2. It would be nice to resolve the cast from the container so we can auto-inject dependencies through their constructors. I don't know the performance impacts on resolving from the container, but if we cache the resolved instances statically I guess the impact would be low.

Reply:
1. Instances are stored in the `castInstances` variable of the model (already renamed to camelCase), so they will be saved between models.
2. You think it would be better to bring the code to the form:
```php
use Illuminate\Container\Container;

if (! array_key_exists($key, $this->castsInstances)) {
    $this->castsInstances[$key] = Container::getInstance()
        ->make($this->getCast($key));
}

return $this->castsInstances[$key];
```
?

Reply:
If I understand correctly, then I consider it to be wrong to turn methods into static ones, since in recent [changes](https://github.com/laravel/framework/pull/30958/commits/ce3985ac7c478bc5f2699a66e2ef1e4feeef06d4) storage of field names and original values has been added, which cannot be obtained with static methods.

Reply:
I think the last comment does not apply anymore after the recent changes, we're back were we were with slightly different naming and no new trait.

From https://github.com/laravel/framework/pull/30958#discussion_r361830332 :
> Regarding memory consumption, would it be a good idea to keep the cast instances in a static array on the model? So the resolved casts are cached between model instances? Something like what is done to the global scopes.

@rodrigopedra very good point!

I think this should be addressed, effectively the custom casts are globally cached singletons.

> It would be nice to resolve the cast from the container so we can auto-inject dependencies through their constructors

I was the one proposing not to use the container make. But this was when the implementation looked differently.

With the suggestion to use single cache for all the castables, I think we can "go back" to use the container because effectively for any given request, a castable is only instantiated once then, which doesn't make it any worse/better then any other (deferred) singleton in use for any given request.

TL;DR: 👍 for a) global (not per-model) castable cache and b) back to container make.

Reply:
Solved

Reply:
I don't think so. You switched back to `Container::make()` but:
- castsInstances is totally unused (performance)
- it's still a per-model cache (memory)

Basically at the moment, the worst of all possibilities 😄

Reply:
The language barrier. I did as I understood. 🙂

Reply:
From all that I understood, I see three options:

1.
```php
if (! array_key_exists($key, $this->castsInstances)) {
    $this->castsInstances[$key] = Container::getInstance()
        ->make($this->getCast($key));
}

return $this->castsInstances[$key];
```

2.
```php
if (! array_key_exists($key, $this->castsInstances)) {
    $cast = $this->getCast($key);

    $this->castsInstances[$key] =new $cast;
}

return $this->castsInstances[$key];
```

A lot of text was written in the comments above and Google Translate is probably incorrect and does not fully translate. 😕

3.
```php
return Container::getInstance()
        ->make($this->getCast($key));
```

Reply:
Do I understand correctly what needs to be done in the likeness of `HasGlobalScopes` trait?

Reply:
If this time I understood correctly, then: https://github.com/laravel/framework/pull/30958/commits/d62cd8d036fe5b5e10be64ab9603601ac0176460#diff-6a9cb621261ef4a702081454957adfb2R1036

Reply:
Continuing my feedback here on the new code part directly => https://github.com/laravel/framework/pull/30958#discussion_r361934723

Reply:
Yes, I already saw it. Thanks!